docs/adrs/00008-re-process-documents.md

```diff
@@ -0,0 +1,103 @@
+Date: 2025-08-08
+
+## Status
+
+DRAFT
+
+## Context
+
+During the process of ingestion, we extract certain information of the uploaded documents and store that information
+in the database. We also store the original source document "as-is".
+
+When making changes to the database structure, we also have a migration process, which takes care of upgrading the
+database structures during an upgrade.
+
+However, in some cases, changing the database structure actually means extracting more information from documents than
+is currently stored in the database. Or information is extracted in a different way. This requires a re-processing of
+all documents affected by this change.
+
+### Example
+
+We do ignore all CVSS v2 scores at the moment. Adding new fields for storing v2 scores, we wouldn't have
+any stored in the database without re-processing documents and extracting that information.
+
+### Assumptions
+
+This ADR makes the following assumptions:
+
+* All documents are stored in the storage
+* It is expected that an upgrade is actually required
+* Running such migrations is expected to take a long time
+* The management of infrastructure (PostgreSQL) is not in the scope of Trustify
+
+Question? Do we want to support downgrades?
+
+## Decision
+
+During the migration of database structures (sea orm), we also re-process all documents (if required). This would
+be running during the migration job of the Helm chart and would have an impact on updates as the rollout of newer
+version pods would be delayed until the migration (of data) has been finished.
+
+This would also require to prevent users from creating new documents during that time. Otherwise, we would need to
+re-process documents ingested during the migration time. A way of doing this could be to leverage PostgreSQL's ability
+to switch into read-only mode. Having mutable operations fail with a 503 (Service Unavailable) error. This would also
+allow for easy A/B (green/blue) database setups. Switching the main one to read-only, having the other one run the
+migration.
+
+We could provide an endpoint to the UI, reporting the fact that the system is in read-only mode during a migration.
+
+* 👍 Can fully migrate database (create mandatory field as optional -> re-process -> make mandatory)
+* 👍 Might allow for an out-of-band migration of data, before running the upgrade (even on a staging env)
+* 👍 Would allow to continue serving data while the process is running
+* 👎 Might be tricky to create a combined re-processing of multiple ones
+* 👎 Might block an upgrade if re-processing fails
+
+### Approach 1
+
+The "lazy" approach, where the user just runs the migration (or the new version of the application with migrations
+enabled). The process will migrate schema and data. This might block the startup for a bit. But would be fast and
+simple for small systems.
+
+### Approach 2
+
+The user uses a green/blue deployment. Switching the application to use green and run migrations against blue. Once
```

Comment:
worth mentioning that amazon has blue/green pg deployments as part of their service though I do not have much experience with them

Reply:
Yes, that was my though here. I'd expect other PG services to have similar setups available. However, it's up to the user to provide a database. We'd just try to ensure we can work with a model like this.